### PR TITLE
deps: Bump to OpenDAL v0.4

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR will bump OpenDAL to v0.4, read [v0.4.0 release note](https://github.com/datafuselabs/opendal/discussions/210) to know more.

For databend, this PR will address the following problems:

- Fix https://github.com/datafuselabs/databend/issues/4582
- Improve error handling of OpenDAL

## Changelog

- Bug Fix
- Improvement

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/exception/src/exception_code.rs [hunk 1]
```diff
@@ -195,13 +195,9 @@ build_exceptions! {
 
 // Storage errors [3001, 4000].
 build_exceptions! {
-    UnknownStorageSchemeName(3001),
-    SecretKeyNotSet(3002),
-    DalTransportError(3003),
-    DalPathNotFound(3004),
-    SerdeError(3005),
-    DalError(3006),
-    DalStatError(3007),
```
Comment (BohuTANG) [inline common/exception/src/exception_code.rs:204]: 👍

File: query/src/storages/s3/s3_stage_source.rs [hunk 2]
```diff
@@ -149,19 +151,24 @@ impl StageSource {
...
+                self.schema.clone(),
+                stage,
+                Box::new(object.reader().await?),
+            )
+            .await?),
+            StageFileFormatType::Parquet => Ok(Self::parquet_source(
+                ctx.clone(),
+                self.schema.clone(),
+                stage,
+                seekable_read(&object, ..),
```
Comment (BohuTANG) [inline query/src/storages/s3/s3_stage_source.rs:169]: Here is not friendly to use, why not `object.seek_reader()`?

Comment (Xuanwo) [reply query/src/storages/s3/s3_stage_source.rs:169]: I want to indicate that it's not a zero-cost operation. But I think it's better to have a `object.seekable_reader()`, let's do this.

**Comment by Xuanwo** [comment]
In this PR, I also fix some clippy warnings around benches.

The most interesting one:

```
error: `assert!(false)` should probably be replaced
  --> query/benches/suites/mod.rs:41:9
   |
41 |         assert!(false)
   |         ^^^^^^^^^^^^^^
   |
   = note: `-D clippy::assertions-on-constants` implied by `-D warnings`
   = help: use `panic!()` or `unreachable!()`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assertions_on_constants
```

It seems our CI doesn't cover benches code. Ask @everpcpc for verification.

**Comment by everpcpc** [comment]
> In this PR, I also fix some clippy warnings around benches.
> 
> The most interesting one:
> 
> ```
> error: `assert!(false)` should probably be replaced
>   --> query/benches/suites/mod.rs:41:9
>    |
> 41 |         assert!(false)
>    |         ^^^^^^^^^^^^^^
>    |
>    = note: `-D clippy::assertions-on-constants` implied by `-D warnings`
>    = help: use `panic!()` or `unreachable!()`
>    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assertions_on_constants
> ```
> 
> It seems our CI doesn't cover benches code. Ask @everpcpc for verification.
> 
> Locally, I use `cargo clippy --all-targets -- -D warnings`

yes, we only run `cargo clippy --tests -- -D warnings` in ci now.
seems came from https://github.com/datafuselabs/databend/commit/e264ce4500cefa42b70b47395c6c7ae8a464d65e#diff-99b87178e201388c0ebfd7d8584c9f6930100d335919c10b7399b39a88cb0c03R27

we'd better change to `--all-targets`

**Comment by BohuTANG** [comment]
Checking one question with this PR:

```
ERROR 1105 (HY000): Code: 3003, displayText = object permission denied: (op: stat, path: fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321, source: response part: Parts { status: 403, version: HTTP/1.1, headers: {"x-amz-request-id": "H7JES5RX8545F3YH", "x-amz-id-2": "aSmaaQtvISxLeRmeb+4cepo3UUk1TbO0anxBdLF5wd9smLB7ANxO7ZcSa6xMe/4KPO94OamIjZQ=", "content-type": "application/xml", "date": "Sun, 03 Apr 2022 06:51:16 GMT", "server": "AmazonS3"} }, body: "").
```

The bucket path:
```
fuse/1 -- exists
fuse/1/5 -- not exists
```

Why the response status is `403` not `404`?
https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#RESTErrorResponses

**Comment by BohuTANG** [comment]
Client:
```
mysql> select version();
+----------------------------------------------------------------------------------------------------+
| version()                                                                                          |
+----------------------------------------------------------------------------------------------------+
| DatabendQuery v0.7.5-nightly-f972974-simd(rust-1.61.0-nightly-2022-04-03T06:15:12.697451893+00:00) |
+----------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
Read 1 rows, 1 B in 0.005 sec., 197.8 rows/sec., 197.8 B/sec.

mysql> select count() from ontime_not_null;                                                                                                                              ERROR 1105 (HY000): Code: 3002, displayText = op: stat, path: fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321, source: response part: Parts { status: 403, version: HTTP/1.1, headers: {"x-amz-request-id": "8VPEPRCGVAXTQ69D", "x-amz-id-2": "Qk37WcBdiP8v7uNI+DYapaaFojsG//WJNTi/1u6p29qL8cWOTly1MV0sD929loZ3lyc1MpbETHs=", "content-type": "application/xml", "date": "Sun, 03 Apr 2022 07:16:46 GMT", "server": "AmazonS3"} }, body: "".

```

Server  error log:
```
^[[2m2022-04-03T07:16:46.569495Z^[[0m ^[[31mERROR^[[0m ^[[2mdatabend_query::servers::mysql::writers::query_result_writer^[[0m^[[2m:^[[0m OnQuery Error: Code: 3002, displayText = op: stat, path: fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321, source: response part: Parts { status: 403, version: HTTP/1.1, headers: {"x-amz-request-id": "8VPEPRCGVAXTQ69D", "x-amz-id-2": "Qk37WcBdiP8v7uNI+DYapaaFojsG//WJNTi/1u6p29qL8cWOTly1MV0sD929loZ3lyc1MpbETHs=", "content-type": "application/xml", "date": "Sun, 03 Apr 2022 07:16:46 GMT", "server": "AmazonS3"} }, body: "".

   0: common_exception::exception_code::<impl common_exception::exception::ErrorCode>::StoragePermissionDenied
             at /home/ubuntu/bohu/github/databend/common/exception/src/exception_code.rs:36:66
      common_exception::exception_into::<impl core::convert::From<std::io::error::Error> for common_exception::exception::ErrorCode>::from
             at /home/ubuntu/bohu/github/databend/common/exception/src/exception_into.rs:111:44
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/result.rs:2064:27
      <&databend_query::sessions::query_ctx::QueryContext as databend_query::storages::fuse::io::read::meta_readers::BufReaderProvider>::buf_reader::{{closure}}
             at /home/ubuntu/bohu/github/databend/query/src/storages/fuse/io/read/meta_readers.rs:139:28
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/mod.rs:91:19
   2: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/future.rs:124:9
      databend_query::storages::fuse::io::read::meta_readers::<impl databend_query::storages::fuse::io::read::cached_reader::Loader<databend_query::storages::fuse::meta::v1::snapshot::TableSnapshot> for T>::load::{{closure}}
             at /home/ubuntu/bohu/github/databend/query/src/storages/fuse/io/read/meta_readers.rs:114:55
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/mod.rs:91:19
   3: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/future.rs:124:9
      databend_query::storages::fuse::io::read::cached_reader::CachedReader<T,L>::load::{{closure}}
             at /home/ubuntu/bohu/github/databend/query/src/storages/fuse/io/read/cached_reader.rs:98:59
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/mod.rs:91:19
   4: databend_query::storages::fuse::io::read::cached_reader::CachedReader<T,L>::read::{{closure}}
             at /home/ubuntu/bohu/github/databend/query/src/storages/fuse/io/read/cached_reader.rs:62:68
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/mod.rs:91:19
      databend_query::storages::fuse::fuse_table::FuseTable::read_table_snapshot::{{closure}}::{{closure}}
             at /home/ubuntu/bohu/github/databend/query/src/storages/fuse/fuse_table.rs:202:57
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/future/mod.rs:91:19
   5: databend_query::storages::fuse::fuse_table::FuseTable::read_table_snapshot::{{closure}}

```

**Comment by Xuanwo** [comment]
> ERROR 1105 (HY000): Code: 3002, displayText = op: stat, path: fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321, source: response part: Parts { status: 403, version: HTTP/1.1, headers: {"x-amz-request-id": "8VPEPRCGVAXTQ69D", "x-amz-id-2": "Qk37WcBdiP8v7uNI+DYapaaFojsG//WJNTi/1u6p29qL8cWOTly1MV0sD929loZ3lyc1MpbETHs=", "content-type": "application/xml", "date": "Sun, 03 Apr 2022 07:16:46 GMT", "server": "AmazonS3"} }, body: "".

Oh, the error message doesn't look nice, we need it print out message like "object permission denied", I will fix it.

---

And to this error, can you try using the same credential to send a HEAD request to the same object (take version/tenant prefix into consideration) `fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321`?

For example:

```shell
aws s3api head-object --bucket=bucket --key=prefix/to/fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321
```

- If `aws s3api` returns 200 or 404, opendal is buggy, and I'm willing to fix it.
- If `aws s3api` returns 403 too, that means maybe we have a bucket policy here which forbid the requests.

**Comment by BohuTANG** [comment]
> > ERROR 1105 (HY000): Code: 3002, displayText = op: stat, path: fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321, source: response part: Parts { status: 403, version: HTTP/1.1, headers: {"x-amz-request-id": "8VPEPRCGVAXTQ69D", "x-amz-id-2": "Qk37WcBdiP8v7uNI+DYapaaFojsG//WJNTi/1u6p29qL8cWOTly1MV0sD929loZ3lyc1MpbETHs=", "content-type": "application/xml", "date": "Sun, 03 Apr 2022 07:16:46 GMT", "server": "AmazonS3"} }, body: "".
> 
> Oh, the error message doesn't look nice, we need it print out message like "object permission denied", I will fix it.
> 
> And to this error, can you try using the same credential to send a HEAD request to the same object (take version/tenant prefix into consideration) `fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321`?
> 
> For example:
> 
> ```shell
> aws s3api head-object --bucket=bucket --key=prefix/to/fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321
> ```
> 
> * If `aws s3api` returns 200 or 404, opendal is buggy, and I'm willing to fix it.
> * If `aws s3api` returns 403 too, that means maybe we have a bucket policy here which forbid the requests.

Thank you for your useful information, have a try and response is 403:
```
aws s3api head-object --bucket=bucket --key=prefix/to/fuse/1/5/_ss/8878969040bd4d7aa5d28297cff1f321

An error occurred (403) when calling the HeadObject operation: Forbidden
```

**Comment by Xuanwo** [comment]
> Unknown output type: s3

Do you have set output type for aws? Please checkout your ` ~/.aws/config`.

With the same command:

```shell
:( aws s3api head-object --bucket=databend-shared --key=fuse/1/1/_ss/92318ff55f694b51b59f119a7c39eda8_v1.json

{
    "AcceptRanges": "bytes",
    "LastModified": "Wed, 30 Mar 2022 07:36:55 GMT",
    "ContentLength": 19254,
    "ETag": "\"d1308468d48c4d8e5572e4f66f3665b8\"",
    "VersionId": "KtVbGRhs8apEOin2ZNsa5T44bciFfupD",
    "ContentType": "binary/octet-stream",
    "Metadata": {}
}
:) aws s3api head-object --bucket=databend-shared --key=fuse/1/1/_ss/92318ff55f694b51b59f119a7c39eda8_v1.jsonx


An error occurred (403) when calling the HeadObject operation: Forbidden
```

**Comment by Xuanwo** [comment]
For a quick workaround, specify `output` in your command:

```shell
aws s3api head-object --bucket=databend-shared --key=fuse/1/1/_ss/92318ff55f694b51b59f119a7c39eda8_v1.json --output=json
```

**Comment by BohuTANG** [comment]
Works:
```
aws s3api head-object --bucket=databend-shared --key=fuse/1/2/_ss/095e11cc31e34a818c36479099308d06_v1.json --output=json
{
    "AcceptRanges": "bytes",
    "LastModified": "Wed, 30 Mar 2022 12:55:56 GMT",
    "ContentLength": 49208,
    "ETag": "\"d424a5210da9a49419f308b2217236e8\"",
    "VersionId": "_4e2wY5ewTUNddKhh2WaKvHaonUyBxQh",
    "ContentType": "binary/octet-stream",
    "Metadata": {}
}

```

The not exists path head-object  returns 403:
```
aws s3api head-object --bucket=databend-shared --key=fuse/1/2non/_ss/095e11cc31e34a818c36479099308d06_v1.json --output=json

An error occurred (403) when calling the HeadObject operation: Forbidden
```

**Comment by Xuanwo** [comment]
It seems an expected behavior of AWS S3: https://stackoverflow.com/questions/19037664/how-do-i-have-an-s3-bucket-return-404-instead-of-403-for-a-key-that-does-not-e

> S3 returns a 403 instead of a 404 when the user doesn't have permission to list the bucket contents.

> If you query for an object and receive a 404, then you know that object doesn't exist. This is information you shouldn't know if you don't have permission to list the bucket contents so instead of telling you it doesn't exist, S3 just tells you that you're trying to do something you're not allowed to do.

**Review by BohuTANG — APPROVED** [review]
looks great to me

**Pull request merged**